railgun/internal/ctrlutils/ingress-status.go

```diff
@@ -0,0 +1,388 @@
+package ctrlutils
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/go-logr/logr"
+	"github.com/kong/deck/file"
+	"github.com/kong/kubernetes-ingress-controller/pkg/sendconfig"
+	kicclientset "github.com/kong/kubernetes-ingress-controller/railgun/pkg/clientset"
+	"github.com/prometheus/common/log"
+	apiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	knative "knative.dev/networking/pkg/apis/networking/v1alpha1"
+	knativeversioned "knative.dev/networking/pkg/client/clientset/versioned"
+	knativeApis "knative.dev/pkg/apis"
+	"knative.dev/pkg/network"
+)
+
+// dedicated function that process ingress/customer resource status update after configuration is updated within kong.
+func PullConfigUpdate(kongConfig sendconfig.Kong, log logr.Logger, ctx context.Context, kubeConfig *rest.Config, stopCh <-chan struct{}) {
+	log.Info("Launching Ingress Status Update Thread.")
+	var wg sync.WaitGroup
+	for {
+		select {
+		case updateDone := <-kongConfig.ConfigDone:
+			log.Info("receive configuration information. Update ingress status %v \n", updateDone)
+			wg.Add(1)
+			go UpdateIngress(&updateDone, log, ctx, kubeConfig, &wg)
```

Comment:
It occurs to me that we will have a race condition bug if we use a goroutine here to perform the status updates: given that there are no guarantees about the order routines will be executed in by the runtime, we could theoretically receive 2 or more updates to `kongConfig.ConfigDone` within a brief millisecond period which would run in _any order_ according to the Go runtime's scheduler NOT according to the order the runtime received them. As such we would have a bug where status updates could be out of order when updates happen rapidly.

My suggestion would be that we drop the goroutine altogether and make this update a blocking call and not process further updates to the kongconfig until we've finished processing the current one.

Reply:
Alternatively: I suppose we could rely on _timestamps_ in order to to make ordering decisions but I'm not exactly sure yet how that would look as far as how we track those timestamps?

Reply:
I dont think we need grant updating sequence. 

The job of each goroutine is to ensure that its updates the assigned ingress which is included within the updateDone - ingress name and namespace, that is unique in the cluster. 

This mechanism resolves concurrency updates ideally.

Reply:
Let me describe the failure scenario, and see if this helps:

if there are two **linear and consecutive** updates to `kongConfig.ConfigDone` we **lose any guarantees of linearity** by firing off the status updates for them in separate goroutines. This presents a technical race condition, no? If I'm understanding this right, you could literally end up with the wrong status for an object.

Reply:
Also I should point out, because we're doing this right now:
```
wg.Add(1)
go UpdateIngress()
wg.Wait()
```

We are _technically already ensuring that we perform status updates in the same order we receive updates to `kongConfig.ConfigDone`.

The only thing is, that if we want to keep that behavior, using a waitgroup is ultimately superfluous, as pointed out in other comments.

Reply:
This was discussed offline and closes w/ follow up.